java-checks/src/main/java/org/sonar/java/checks/MethodParametersOrderCheck.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S2234")
+public class MethodParametersOrderCheck extends IssuableSubscriptionVisitor {
+
+  private Map<Symbol, LinkedHashSet<String>> methodNameKeyAndParameteresValue = new HashMap<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    methodNameKeyAndParameteresValue.clear();
+    super.scanFile(context);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD)) {
+      MethodTree methodTree = (MethodTree) tree;
+      methodNameKeyAndParameteresValue.put(methodTree.symbol(), new LinkedHashSet<>(methodTree.parameters().stream().map(param -> param.simpleName().name()
+        .toLowerCase(Locale.ENGLISH)).collect(Collectors.toList())));
+      return;
+    } else {
+      MethodInvocationTree methodInvTree = (MethodInvocationTree) tree;
+      LinkedHashSet<String> definedFunctionParameters = methodNameKeyAndParameteresValue.get(methodInvTree.symbol());
+      LinkedHashSet<String> invocationParameters = new LinkedHashSet<>(
```

Comment:
You can probably use `Set` directly here, and previous variable as well.

Reply:
Yes, but we try to limit as much as possible use of implementation class in variable declaration in Java.
Using interfaces rather than implementation classes reduces coupling and allow more flexibility. In this particular case, using a `LinkedHashSet` is an implementation details (maintaining the order in your `set`), but you are not using any particular method which would be defined only in `LinkedHashSet` and not in `Set`. Refer to this article: https://www.javaworld.com/article/2073649/core-java/why-extends-is-evil.html 

It should be:
```Set<String> invocationParameters = new LinkedHashSet<>(... ``` 
instead of 
```LinkedHashSet<String> invocationParameters = new LinkedHashSet<>(... ```

Reply:
And we should use Lists.